### ISSUE TITLE
idea: gravitational pull (weapons: tractor beam, stellar objects: wormholes) 

### ISSUE THREAD

**Issue by nylki** [endless-sky/endless-sky#912]
What about having `StellarObjects` that have a gravitational pull for example? Some interesting phenomena would be possible, like black holes or wormholes that pull ships closers and eventually damage ships that get too close
Or weapons that pull lighter ships closer. Useful for advanced (pirate) ships perhaps, that have short range weapons and want to pull slow ships to them.

I was thinking how to best implement them.
For weapons: Adding a new property called `gravitationalPull`, that gets checked on `Ship::TakeDamage()` and calculated based it's value, the mass of the attacked ship and the distance of the attacking ship to the attacked. I think I'll just try it out in a fork.

Not sure about stellar objects though. In a perfect world, asteroids would have to be affected as wel :)

What do you think?

**Comment by jafdy** [comment]
With weapons i think you can use a negative `hit force`.

**Comment by nylki** [comment]
@jafdy yeah, I realized that myself shortly after posting the issue. see my edit :)

**Comment by endless-sky** [comment]
Are you volunteering to modify the AI to be able to plot flight paths under the effects of gravity? :)

In Escape Velocity (the game ES is patterned after) gravity existed in a few systems, but it was really annoying and poorly supported by the AI.

**Comment by endless-sky** [comment]
Also, note that if gravity were added, it would be possible for slow ships to be unable to escape the system center without complicated gravitational slingshot maneuvers. Other than that, though, it might be fun to add just to see what flying with gravity would feel like...

**Comment by nylki** [comment]
> In Escape Velocity (the game ES is patterned after) gravity existed in a few systems, but it was really annoying and poorly supported by the AI.

I remember that! You are right, smaller ships got often caught in the gravitational pull.
First of all, big gravitational effects should only be special cases. We could cheat in those situation, and let ships have always a minimum positive velocity when using their engines, so small ships could always escape although slowly (this would still make it interesting because the ship is in a disadvantage). While not using engines, the gravitation could have it's un-cheated effect.

> Are you volunteering to modify the AI to be able to plot flight paths under the effects of gravity? :)

Hehe, I did not think of the implications if we added it to regular stellar objects like stars or planets, but only to a few selected special objects like wormholes. It could get tricky to do the necessary calculations for the AI to handle situations properly (eg. fly around huge wormholes if they don't want to fly to its center).

I would be willing to atleast try things out in a separate fork, implementing first some basic gravity, applying it to a few test systems and then testing for some AI changes. Just for fun :) If it is fun to play and the AI can handle it, I'd make a PR.

**Comment by nylki** [comment]
@endless-sky 
I report success!
Played around with some newtonian gravity.  So far I only used the radius of stellar objects to calculate the mass, of course thats not so realistic for stars ;)

Here is a video with strong gravity, that shows some interesting slingshot physics:
[![](https://i.ytimg.com/vi/vDNXytSEYSw/2.jpg)](https://youtu.be/vDNXytSEYSw)

And here is a video where the gravity makes more sense gameplay wise and is actually quite fun.
I also tested a tractor beam. I can think of some fun missions, where you have to rescue some derelict shuttles from high gravity areas :)
[![](https://i.ytimg.com/vi/sW5lP__lsgc/3.jpg)](https://youtu.be/sW5lP__lsgc)

Modifying the AI to properly work with gravity is a completely different story of course. In most cases it works surprisingly well with a lower gravitational constant. The main problem I noticed is ships failing to adjust their course to go into hyperspace. Which makes them wobble.

Next thing I might try is using some flow field based path finding to calculate way points when a ship wants to fly to a certain coordinate.

**Comment by nylki** [comment]
New update:

- Stars have much more mass than planets now
- Mass is configurable per StellarObject in `maps.txt` via mass` keyword. If it is not set, the mass gets calculated via the `radius` and if it is a star
- Ships can also take heat damage and regular damage when too close to certain objects (also configurable via `maps.txt`). Stars creat heat and regular damage by default.

**Comment by jafdy** [comment]
I don't think that it is a good idea to have all stars do damage as it may be to much for a new player. If you are moving for no apparent reason (you don't know gravity applys) then you get near a star and your ship starts taking damage with no enemy in sight. It may discourage some people.

**Comment by nylki** [comment]
@jafdy Well new players would very quickly learn about this and adapt their flight paths in the future. ES is not a permadeath game after all.

But I agree with you, that a warning would still be useful in that case. And damage from system object should probably be restricted to some special cases. I was mainly playing around a bit and see if my changes had an effect. :)

Btw. have you played around with the gravity fork a bit?

**Comment by jafdy** [comment]
No, I never could work out how most of GitHub works.

**Comment by Bladewood** [comment]
This is going to screw with my ramscoop strategy :S
